implementations/rust/ockam/ockam/src/entity/profile/profile_change_type/rotate_key.rs

```diff
@@ -0,0 +1,106 @@
+use crate::entity::profile::event_identifier::EventIdentifier;
+use crate::entity::profile::profile_change::{ProfileChange, ProfileKeyPurpose, ProfileKeyType};
+use crate::entity::profile::profile_change_type::ProfileChangeType;
+use crate::entity::profile::proof::{ProfileChangeProof, Signature, SignatureType};
+use crate::entity::{Changes, Profile, ProfileChangeEvent, ProfileEventAttributes};
+use crate::OckamError;
+use ockam_vault_core::types::{SecretAttributes, SecretPersistence, SecretType};
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct RotateKeyChange {
+    label: String,
+    key_type: ProfileKeyType,
+    key_purpose: ProfileKeyPurpose,
+    public_key: Vec<u8>,
+}
+
+impl RotateKeyChange {
+    pub fn label(&self) -> &str {
+        &self.label
+    }
+    pub fn key_type(&self) -> ProfileKeyType {
+        self.key_type
+    }
+    pub fn key_purpose(&self) -> ProfileKeyPurpose {
+        self.key_purpose
+    }
+    pub fn public_key(&self) -> &[u8] {
+        &self.public_key
+    }
+}
+
+impl RotateKeyChange {
+    pub fn new(
+        label: String,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+        public_key: Vec<u8>,
+    ) -> Self {
+        RotateKeyChange {
+            label,
+            key_type,
+            key_purpose,
+            public_key,
+        }
+    }
+}
+
+impl Profile {
+    pub(crate) fn rotate_key_event(
+        &self,
+        label: String,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+        attributes: Option<ProfileEventAttributes>,
+    ) -> Result<ProfileChangeEvent, ockam_core::Error> {
+        let attributes = attributes.unwrap_or(ProfileEventAttributes::new());
+
+        let prev_event_id = self.get_last_event_id()?;
+
+        let last_event_in_chain = self.find_last_key_event(&label, key_type, key_purpose)?;
+
+        let last_key_in_chain =
+            self.get_private_key_from_event(&label, key_type, key_purpose, last_event_in_chain)?;
+
+        let mut v = self.vault.lock().unwrap();
+
+        // TODO: Should be customisable
+        let secret_attributes = SecretAttributes {
+            stype: SecretType::Curve25519,
+            persistence: SecretPersistence::Persistent,
+            length: 0,
+        };
+
+        let private_key = v.secret_generate(secret_attributes)?;
+        let public_key = v.secret_public_key_get(&private_key)?.as_ref().to_vec();
+
+        let event = RotateKeyChange::new(label, key_type, key_purpose, public_key);
+
+        let change = ProfileChange::new(
+            1,
```

Comment:
is this hardcoded to 1 or does it change according to how many events have already happened?

Reply:
it's version number, still needs to be a named const...

Reply:
Thanks for the clarification.